pkg/core/resources/apis/meshservice/generate/generator.go

```diff
@@ -175,22 +180,51 @@ func (g *Generator) generate(ctx context.Context, mesh string, dataplanes []*cor
 			log.Info("Port conflict for a kuma.io/service tag, ports must be identical across Dataplane inbounds for a given kuma.io/service", "dps", dps)
 		}
 		delete(meshservicesByName, meshService.GetMeta().GetName())
-		if newMeshService != nil && servicesDiffer(meshService.Spec, newMeshService) {
+		gracePeriodStartedAtText, hasGracePeriodLabel := meshService.GetMeta().GetLabels()[gracePeriodStartedLabel]
+
+		if newMeshService != nil && (servicesDiffer(meshService.Spec, newMeshService) || hasGracePeriodLabel) {
 			meta := meshService.GetMeta()
 			meshService = meshservice_api.NewMeshServiceResource()
 			meshService.Meta = meta
 			meshService.Spec = newMeshService
-			if err := g.resManager.Update(ctx, meshService); err != nil {
+
+			// Unset the grace period by deleting the label
+			newLabels := maps.Clone(meshService.GetMeta().GetLabels())
+			delete(newLabels, gracePeriodStartedLabel)
+
+			if err := g.resManager.Update(ctx, meshService, store.UpdateWithLabels(newLabels)); err != nil {
 				log.Error(err, "couldn't update MeshService")
 				continue
 			}
 			log.Info("updated MeshService")
 		} else if newMeshService == nil {
-			if err := g.resManager.Delete(ctx, meshservice_api.NewMeshServiceResource(), store.DeleteBy(model.MetaToResourceKey(meshService.GetMeta()))); err != nil {
-				log.Error(err, "couldn't delete MeshService")
-				continue
+			gracePeriodStartedAt, err := time.Parse(time.RFC3339, gracePeriodStartedAtText)
+			if hasGracePeriodLabel && err == nil {
+				// If we have a valid grace period set, check if it's expired
+				if time.Since(gracePeriodStartedAt) > g.gracePeriodInterval {
+					if err := g.resManager.Delete(ctx, meshservice_api.NewMeshServiceResource(), store.DeleteBy(model.MetaToResourceKey(meshService.GetMeta()))); err != nil {
+						log.Error(err, "couldn't delete MeshService")
+						continue
+					}
+					log.Info("deleted MeshService")
+				}
+			} else {
+				// Start the grace period if we don't have the label or it's invalid
+				if hasGracePeriodLabel && err != nil {
+					log.Info("couldn't parse grace period label, ignoring", "value", gracePeriodStartedAtText)
+				}
+				nowText, err := time.Now().MarshalText()
+				if err != nil {
+					log.Error(err, "couldn't marshal time.Now as text, this shouldn't be possible")
+					continue
+				}
+				newLabels := maps.Clone(meshService.GetMeta().GetLabels())
+				newLabels[gracePeriodStartedLabel] = string(nowText)
+				if err := g.resManager.Update(ctx, meshService, store.UpdateWithLabels(newLabels)); err != nil {
+					log.Error(err, "couldn't update MeshService")
+					continue
+				}
```

Comment:
Let's log something like `log.Info("mesh service marked as ready to delete", "deleteIn", gracePeriod)`. `"deleteIn"` sounds stupid, there is probably a better word for it

Reply:
done